sdlib/d/gc/thread.d

```diff
@@ -88,14 +88,29 @@ void restartTheWorld() {
 	gThreadState.restartTheWorld();
 }
 
-void threadScan(ScanDg scan) {
+void threadScan(ThreadCache* tc, ScanDg scan) {
 	// Scan the registered TLS segments.
-	foreach (s; threadCache.tlsSegments) {
+	foreach (s; tc.tlsSegments) {
 		scan(s);
 	}
 
-	import d.gc.stack;
-	scanStack(scan);
+	// Skip scanning the stack if someone else is handling it.
+	if (tc.stackTop == SkipScanStack) {
+		return;
+	}
+
+	// Scan a valid stack as described by the threadCache.
+	if (tc.stackTop > SkipScanStack) {
+		import d.gc.range;
+		scan(makeRange(tc.stackTop, tc.stackBottom));
+		return;
+	}
+
+	// Scan our own stack with registers if requested.
+	if (tc is &threadCache) {
+		import d.gc.stack;
+		scanStack(scan);
+	}
```

Comment:
This is not good. The caller knows which codepath is going to be taken, because it. This is better than the previous iteration, but it should all sink to the same place, not forget and then reconstruct knowledge by doing tests that someone will end up breaking down the road.